yjit/src/stats.rs

```diff
@@ -650,8 +652,21 @@ pub extern "C" fn rb_yjit_print_stats_p(_ec: EcPtr, _ruby_self: VALUE) -> VALUE
 /// Primitive called in yjit.rb.
 /// Export all YJIT statistics as a Ruby hash.
 #[no_mangle]
-pub extern "C" fn rb_yjit_get_stats(_ec: EcPtr, _ruby_self: VALUE) -> VALUE {
-    with_vm_lock(src_loc!(), || rb_yjit_gen_stats_dict())
+pub extern "C" fn rb_yjit_get_stats(_ec: EcPtr, _ruby_self: VALUE, hash_or_key: VALUE) -> VALUE {
+    // Do this before taking vm lock.
+    let (hash, key): (VALUE, VALUE) = if unsafe { RB_TYPE_P(hash_or_key, RUBY_T_HASH) } {
+        (hash_or_key, Qnil)
+    } else if unsafe { RB_TYPE_P(hash_or_key, RUBY_T_SYMBOL) } {
+        (Qnil, hash_or_key)
+    } else if hash_or_key.nil_p() {
+        (unsafe { rb_hash_new() }, Qnil)
+    } else {
+        unsafe {
+            rb_raise(rb_eTypeError, b"non-hash or symbol given\0".as_ref().as_ptr() as *const raw::c_char)
```

Comment:
I got this from https://github.com/ruby/ruby/blob/e7fbdf8187ab62bfbe8b2660eda01de94341412e/yjit/src/yjit.rs#L104
but it probably makes sense to build some sort of wrapper to do this.

Reply:
Have you considered splitting this `Primitive` into multiple different `Primitive`s and choosing one in the Ruby method? It might be easier in Ruby than in Rust to match Ruby arguments and raise a Ruby exception while leaving other YJIT-specific stuff to Rust.

Reply:
I was curious about that as I do think it would be simpler.
I wasn't sure if there would be any notable difference between using `is_a?` or `===` and using the `RB_TYPE_P` but if that's not important I will move that there.  Then we don't need to bring in `rb_raise`.

Reply:
Even if you don't split it, I think it would help to reject unexpected cases in Ruby. `YJIT.enable` normalizes the `stats` argument in Ruby first, for example. If we reject all not-Symbol-or-nil cases in Ruby and give `key_or_nil` to the `Primitive`, we could just use a single Primitive.